{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6c06416b-bebd-4cdb-8251-d57ea84291d3",
   "metadata": {},
   "source": [
    "# Task 3\n",
    "## Creating alternative differential expression modules\n",
    "---\n",
    "## Relevance:  \n",
    "While tools like sceptre are very accurated calculating differential perturbation it can take a long time to calculate the differential expression for all genes. \n",
    "Another researchers can have interest in extend the differential expression analysis to tools. The aim of this task is to create a module that can be used to calculate differential expression using a mudata contaning  a mudata object with the informations about guides and scRNA-seq , covariates and coordinate information \n",
    "\n",
    "## Practical description: TODO\n",
    "\n",
    "-  [x] Download the mudata_results.h4mu and transform in a mudata_processed_guides_and_scrna.h5mu by removing unecessary information (code already provided \"Downloading the File and creating mudata with guides and scRNA\")\n",
    "-  [ ] Create the enviroment by running the provided code. Check on  \"Creating the enviroment\" in the last notebook session\n",
    "-  [ ] Create a nextflow process called differential_pertubation that takes as input the mudata_processed_guides_and_scrna.h5mu and a parameter contaning the information about in_trans=TRUE or FALSE.\n",
    "-  [ ] Use the script  differential_pertubation.py to calculate the differential expression.\n",
    "- [ ] return all the result in a dir called  results_dir\n",
    "- [ ] Consider if we need a single file for each diff perturbation tool or if we can receive the current mudata and  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1941d336-865f-4631-90ec-76a380e55019",
   "metadata": {},
   "source": [
    "# Downloading and preprocessing experiment data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0187a36f",
   "metadata": {},
   "source": [
    "## Load MuData containing scRNA data and test results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dd845a12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan\r\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e999e51-f8db-4d53-8c8d-8a04f45b4f29",
   "metadata": {
    "scrolled": false,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-04-03 15:28:35--  https://www.dropbox.com/s/5jysq76bttem3jd/mudata_results.h5mu?dl=1\n",
      "Resolving www.dropbox.com (www.dropbox.com)... 162.125.3.18, 2620:100:6018:18::a27d:312\n",
      "Connecting to www.dropbox.com (www.dropbox.com)|162.125.3.18|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: /s/dl/5jysq76bttem3jd/mudata_results.h5mu [following]\n",
      "--2023-04-03 15:28:36--  https://www.dropbox.com/s/dl/5jysq76bttem3jd/mudata_results.h5mu\n",
      "Reusing existing connection to www.dropbox.com:443.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://ucec8a7756382efe1128dedc366f.dl.dropboxusercontent.com/cd/0/get/B5dB2U89Hr1-yybV1aw4kpfx6ljmnjHBQANXLOePiAikKe9wr1OK7uOl5qizlvuF9t5tMkM34kc1mJofN8nPck-cnWB_Iji1zlhuN_PZsSp0UdlRk21zCb-7xROit4uQgNeGNIaEN8WeBJp32bxeHerUWt_HIrHWt2212roktcbHi22Qq_OvVmvC0H_REg3v9D0/file?dl=1# [following]\n",
      "--2023-04-03 15:28:37--  https://ucec8a7756382efe1128dedc366f.dl.dropboxusercontent.com/cd/0/get/B5dB2U89Hr1-yybV1aw4kpfx6ljmnjHBQANXLOePiAikKe9wr1OK7uOl5qizlvuF9t5tMkM34kc1mJofN8nPck-cnWB_Iji1zlhuN_PZsSp0UdlRk21zCb-7xROit4uQgNeGNIaEN8WeBJp32bxeHerUWt_HIrHWt2212roktcbHi22Qq_OvVmvC0H_REg3v9D0/file?dl=1\n",
      "Resolving ucec8a7756382efe1128dedc366f.dl.dropboxusercontent.com (ucec8a7756382efe1128dedc366f.dl.dropboxusercontent.com)... 162.125.3.15, 2620:100:6018:15::a27d:30f\n",
      "Connecting to ucec8a7756382efe1128dedc366f.dl.dropboxusercontent.com (ucec8a7756382efe1128dedc366f.dl.dropboxusercontent.com)|162.125.3.15|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 146696468 (140M) [application/binary]\n",
      "Saving to: ‘mudata_results.h5mu’\n",
      "\n",
      "mudata_results.h5mu 100%[===================>] 139.90M   102MB/s    in 1.4s    \n",
      "\n",
      "2023-04-03 15:28:39 (102 MB/s) - ‘mudata_results.h5mu’ saved [146696468/146696468]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://www.dropbox.com/s/5jysq76bttem3jd/mudata_results.h5mu?dl=1 -O mudata_results.h5mu"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c589eac",
   "metadata": {},
   "source": [
    "### Inspect combined MuData object containing both raw data and test summary metrics\n",
    "This is the format for data & test results that is currently expected for use in task 4, so our output should look similar to this "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8e44897",
   "metadata": {
    "scrolled": false,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/opt/conda/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/opt/conda/lib/python3.10/site-packages/mudata/_core/mudata.py:458: UserWarning: Cannot join columns with the same name because var_names are intersecting.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 5218 × 17212\n",
       "  4 modalities\n",
       "    guides:\t5109 x 98\n",
       "      obs:\t&#x27;bath_number&#x27;, &#x27;percent_mito&#x27;, &#x27;log_number_of_detected_genes&#x27;, &#x27;log_total_gene_count&#x27;, &#x27;log_total_guide_count&#x27;\n",
       "      var:\t&#x27;feature_name&#x27;, &#x27;guide_chr&#x27;, &#x27;guide_end&#x27;, &#x27;guide_start&#x27;, &#x27;guide_number&#x27;, &#x27;target_elements&#x27;\n",
       "    scRNA:\t5109 x 16030\n",
       "      obs:\t&#x27;bath_number&#x27;, &#x27;percent_mito&#x27;, &#x27;log_number_of_detected_genes&#x27;, &#x27;log_total_gene_count&#x27;, &#x27;log_total_guide_count&#x27;\n",
       "      var:\t&#x27;feature_name&#x27;, &#x27;n_cells&#x27;, &#x27;transcript_chr&#x27;, &#x27;transcript_start&#x27;, &#x27;transcript_end&#x27;, &#x27;ensg&#x27;\n",
       "    result_guides:\t68 x 542\n",
       "      obs:\t&#x27;feature_name&#x27;, &#x27;guide_chr&#x27;, &#x27;guide_end&#x27;, &#x27;guide_start&#x27;, &#x27;guide_number&#x27;, &#x27;target_elements&#x27;\n",
       "      var:\t&#x27;feature_name&#x27;, &#x27;n_cells&#x27;, &#x27;transcript_chr&#x27;, &#x27;transcript_start&#x27;, &#x27;transcript_end&#x27;, &#x27;ensg&#x27;\n",
       "      layers:\t&#x27;adj_pvalue&#x27;, &#x27;log_fold_change&#x27;, &#x27;significant&#x27;, &#x27;z_value&#x27;\n",
       "    result_elements:\t41 x 542\n",
       "      obs:\t&#x27;element&#x27;, &#x27;element_chr&#x27;, &#x27;element_start&#x27;, &#x27;element_end&#x27;\n",
       "      var:\t&#x27;feature_name&#x27;, &#x27;n_cells&#x27;, &#x27;transcript_chr&#x27;, &#x27;transcript_start&#x27;, &#x27;transcript_end&#x27;, &#x27;ensg&#x27;\n",
       "      layers:\t&#x27;sig_not_adj&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 5218 × 17212\n",
       "  4 modalities\n",
       "    guides:\t5109 x 98\n",
       "      obs:\t'bath_number', 'percent_mito', 'log_number_of_detected_genes', 'log_total_gene_count', 'log_total_guide_count'\n",
       "      var:\t'feature_name', 'guide_chr', 'guide_end', 'guide_start', 'guide_number', 'target_elements'\n",
       "    scRNA:\t5109 x 16030\n",
       "      obs:\t'bath_number', 'percent_mito', 'log_number_of_detected_genes', 'log_total_gene_count', 'log_total_guide_count'\n",
       "      var:\t'feature_name', 'n_cells', 'transcript_chr', 'transcript_start', 'transcript_end', 'ensg'\n",
       "    result_guides:\t68 x 542\n",
       "      obs:\t'feature_name', 'guide_chr', 'guide_end', 'guide_start', 'guide_number', 'target_elements'\n",
       "      var:\t'feature_name', 'n_cells', 'transcript_chr', 'transcript_start', 'transcript_end', 'ensg'\n",
       "      layers:\t'adj_pvalue', 'log_fold_change', 'significant', 'z_value'\n",
       "    result_elements:\t41 x 542\n",
       "      obs:\t'element', 'element_chr', 'element_start', 'element_end'\n",
       "      var:\t'feature_name', 'n_cells', 'transcript_chr', 'transcript_start', 'transcript_end', 'ensg'\n",
       "      layers:\t'sig_not_adj'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import mudata as md\n",
    "mu_results = md.read('/home/jovyan/mudata_results.h5mu')\n",
    "mu_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8af035c",
   "metadata": {},
   "source": [
    "### Remove test results from MuData (guides and genes only) and inspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e141ee0c-c0b7-4893-bab7-211c4a82849e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "mu_guides_and_rna = md.MuData({'guides': mu_results['guides'], 'scRNA':mu_results['scRNA']})\n",
    "mu_guides_and_rna.write('mudata_processed_guides_and_scrna.h5mu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e7efe197-db3d-49e4-925c-d1d4c78f7bfd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5109 × 98\n",
       "    obs: 'bath_number', 'percent_mito', 'log_number_of_detected_genes', 'log_total_gene_count', 'log_total_guide_count'\n",
       "    var: 'feature_name', 'guide_chr', 'guide_end', 'guide_start', 'guide_number', 'target_elements'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu_guides_and_rna['guides']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cb863f82-710f-47cc-a706-c64e312ee305",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bath_number</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>log_number_of_detected_genes</th>\n",
       "      <th>log_total_gene_count</th>\n",
       "      <th>log_total_guide_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGGCTT</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.513429</td>\n",
       "      <td>5.517453</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAATACAG</th>\n",
       "      <td>1</td>\n",
       "      <td>0.022816</td>\n",
       "      <td>8.161660</td>\n",
       "      <td>8.161946</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTCGACG</th>\n",
       "      <td>1</td>\n",
       "      <td>0.051611</td>\n",
       "      <td>7.784473</td>\n",
       "      <td>7.784889</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCTTCC</th>\n",
       "      <td>1</td>\n",
       "      <td>0.053218</td>\n",
       "      <td>7.934155</td>\n",
       "      <td>7.934513</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGCCTAA</th>\n",
       "      <td>1</td>\n",
       "      <td>0.045982</td>\n",
       "      <td>8.531490</td>\n",
       "      <td>8.531688</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCTGCAGG</th>\n",
       "      <td>1</td>\n",
       "      <td>0.027573</td>\n",
       "      <td>7.791110</td>\n",
       "      <td>7.791523</td>\n",
       "      <td>2.079442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCGTCCGTT</th>\n",
       "      <td>1</td>\n",
       "      <td>0.027464</td>\n",
       "      <td>8.423102</td>\n",
       "      <td>8.423322</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGCGAGAAA</th>\n",
       "      <td>1</td>\n",
       "      <td>0.040875</td>\n",
       "      <td>7.649693</td>\n",
       "      <td>7.650169</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACATAACC</th>\n",
       "      <td>1</td>\n",
       "      <td>0.031246</td>\n",
       "      <td>8.506739</td>\n",
       "      <td>8.506941</td>\n",
       "      <td>1.386294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACAGGGTACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0.036355</td>\n",
       "      <td>8.281977</td>\n",
       "      <td>8.282230</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5109 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  bath_number  percent_mito  log_number_of_detected_genes  \\\n",
       "AAACCTGAGAGGGCTT            1      0.000000                      5.513429   \n",
       "AAACCTGCAAATACAG            1      0.022816                      8.161660   \n",
       "AAACCTGCACTCGACG            1      0.051611                      7.784473   \n",
       "AAACCTGCAGTCTTCC            1      0.053218                      7.934155   \n",
       "AAACCTGCATGCCTAA            1      0.045982                      8.531490   \n",
       "...                       ...           ...                           ...   \n",
       "TTTGGTTTCCTGCAGG            1      0.027573                      7.791110   \n",
       "TTTGGTTTCGTCCGTT            1      0.027464                      8.423102   \n",
       "TTTGTCAAGCGAGAAA            1      0.040875                      7.649693   \n",
       "TTTGTCACACATAACC            1      0.031246                      8.506739   \n",
       "TTTGTCACAGGGTACA            1      0.036355                      8.281977   \n",
       "\n",
       "                  log_total_gene_count  log_total_guide_count  \n",
       "AAACCTGAGAGGGCTT              5.517453               0.000000  \n",
       "AAACCTGCAAATACAG              8.161946               0.000000  \n",
       "AAACCTGCACTCGACG              7.784889               0.000000  \n",
       "AAACCTGCAGTCTTCC              7.934513               0.693147  \n",
       "AAACCTGCATGCCTAA              8.531688               0.000000  \n",
       "...                                ...                    ...  \n",
       "TTTGGTTTCCTGCAGG              7.791523               2.079442  \n",
       "TTTGGTTTCGTCCGTT              8.423322               0.693147  \n",
       "TTTGTCAAGCGAGAAA              7.650169               0.000000  \n",
       "TTTGTCACACATAACC              8.506941               1.386294  \n",
       "TTTGTCACAGGGTACA              8.282230               0.000000  \n",
       "\n",
       "[5109 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu_guides_and_rna['guides'].obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cb5e89d0-deac-4090-a9e0-73687bef388c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_name</th>\n",
       "      <th>guide_chr</th>\n",
       "      <th>guide_end</th>\n",
       "      <th>guide_start</th>\n",
       "      <th>guide_number</th>\n",
       "      <th>target_elements</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BROX|1_sgrna_chr1:222712600:222712620</th>\n",
       "      <td>BROX|1_sgrna_chr1:222712600:222712620</td>\n",
       "      <td>chr1</td>\n",
       "      <td>222712620</td>\n",
       "      <td>222712600</td>\n",
       "      <td>1</td>\n",
       "      <td>BROX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BROX|2_sgrna_chr1:222712744:222712764</th>\n",
       "      <td>BROX|2_sgrna_chr1:222712744:222712764</td>\n",
       "      <td>chr1</td>\n",
       "      <td>222712764</td>\n",
       "      <td>222712744</td>\n",
       "      <td>2</td>\n",
       "      <td>BROX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCND2|1_sgrna_chr12:4273763:4273783</th>\n",
       "      <td>CCND2|1_sgrna_chr12:4273763:4273783</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4273783</td>\n",
       "      <td>4273763</td>\n",
       "      <td>1</td>\n",
       "      <td>CCND2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCND2|2_sgrna_chr12:4273832:4273852</th>\n",
       "      <td>CCND2|2_sgrna_chr12:4273832:4273852</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4273852</td>\n",
       "      <td>4273832</td>\n",
       "      <td>2</td>\n",
       "      <td>CCND2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEBPB|1_sgrna_chr20:50190830:50190850</th>\n",
       "      <td>CEBPB|1_sgrna_chr20:50190830:50190850</td>\n",
       "      <td>chr20</td>\n",
       "      <td>50190850</td>\n",
       "      <td>50190830</td>\n",
       "      <td>1</td>\n",
       "      <td>CEBPB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UAP1|2_sgrna_chr1:162561996:162562016</th>\n",
       "      <td>UAP1|2_sgrna_chr1:162561996:162562016</td>\n",
       "      <td>chr1</td>\n",
       "      <td>162562016</td>\n",
       "      <td>162561996</td>\n",
       "      <td>2</td>\n",
       "      <td>UAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBE2T|1_sgrna_chr1:202341918:202341938</th>\n",
       "      <td>UBE2T|1_sgrna_chr1:202341918:202341938</td>\n",
       "      <td>chr1</td>\n",
       "      <td>202341938</td>\n",
       "      <td>202341918</td>\n",
       "      <td>1</td>\n",
       "      <td>UBE2T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBE2T|2_sgrna_chr1:202341907:202341927</th>\n",
       "      <td>UBE2T|2_sgrna_chr1:202341907:202341927</td>\n",
       "      <td>chr1</td>\n",
       "      <td>202341927</td>\n",
       "      <td>202341907</td>\n",
       "      <td>2</td>\n",
       "      <td>UBE2T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VDAC1|1_sgrna_chr5:134004687:134004707</th>\n",
       "      <td>VDAC1|1_sgrna_chr5:134004687:134004707</td>\n",
       "      <td>chr5</td>\n",
       "      <td>134004707</td>\n",
       "      <td>134004687</td>\n",
       "      <td>1</td>\n",
       "      <td>VDAC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VDAC1|2_sgrna_chr5:134004700:134004720</th>\n",
       "      <td>VDAC1|2_sgrna_chr5:134004700:134004720</td>\n",
       "      <td>chr5</td>\n",
       "      <td>134004720</td>\n",
       "      <td>134004700</td>\n",
       "      <td>2</td>\n",
       "      <td>VDAC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  feature_name  \\\n",
       "feature_id                                                                       \n",
       "BROX|1_sgrna_chr1:222712600:222712620    BROX|1_sgrna_chr1:222712600:222712620   \n",
       "BROX|2_sgrna_chr1:222712744:222712764    BROX|2_sgrna_chr1:222712744:222712764   \n",
       "CCND2|1_sgrna_chr12:4273763:4273783        CCND2|1_sgrna_chr12:4273763:4273783   \n",
       "CCND2|2_sgrna_chr12:4273832:4273852        CCND2|2_sgrna_chr12:4273832:4273852   \n",
       "CEBPB|1_sgrna_chr20:50190830:50190850    CEBPB|1_sgrna_chr20:50190830:50190850   \n",
       "...                                                                        ...   \n",
       "UAP1|2_sgrna_chr1:162561996:162562016    UAP1|2_sgrna_chr1:162561996:162562016   \n",
       "UBE2T|1_sgrna_chr1:202341918:202341938  UBE2T|1_sgrna_chr1:202341918:202341938   \n",
       "UBE2T|2_sgrna_chr1:202341907:202341927  UBE2T|2_sgrna_chr1:202341907:202341927   \n",
       "VDAC1|1_sgrna_chr5:134004687:134004707  VDAC1|1_sgrna_chr5:134004687:134004707   \n",
       "VDAC1|2_sgrna_chr5:134004700:134004720  VDAC1|2_sgrna_chr5:134004700:134004720   \n",
       "\n",
       "                                       guide_chr  guide_end guide_start  \\\n",
       "feature_id                                                                \n",
       "BROX|1_sgrna_chr1:222712600:222712620       chr1  222712620   222712600   \n",
       "BROX|2_sgrna_chr1:222712744:222712764       chr1  222712764   222712744   \n",
       "CCND2|1_sgrna_chr12:4273763:4273783        chr12    4273783     4273763   \n",
       "CCND2|2_sgrna_chr12:4273832:4273852        chr12    4273852     4273832   \n",
       "CEBPB|1_sgrna_chr20:50190830:50190850      chr20   50190850    50190830   \n",
       "...                                          ...        ...         ...   \n",
       "UAP1|2_sgrna_chr1:162561996:162562016       chr1  162562016   162561996   \n",
       "UBE2T|1_sgrna_chr1:202341918:202341938      chr1  202341938   202341918   \n",
       "UBE2T|2_sgrna_chr1:202341907:202341927      chr1  202341927   202341907   \n",
       "VDAC1|1_sgrna_chr5:134004687:134004707      chr5  134004707   134004687   \n",
       "VDAC1|2_sgrna_chr5:134004700:134004720      chr5  134004720   134004700   \n",
       "\n",
       "                                       guide_number target_elements  \n",
       "feature_id                                                           \n",
       "BROX|1_sgrna_chr1:222712600:222712620             1            BROX  \n",
       "BROX|2_sgrna_chr1:222712744:222712764             2            BROX  \n",
       "CCND2|1_sgrna_chr12:4273763:4273783               1           CCND2  \n",
       "CCND2|2_sgrna_chr12:4273832:4273852               2           CCND2  \n",
       "CEBPB|1_sgrna_chr20:50190830:50190850             1           CEBPB  \n",
       "...                                             ...             ...  \n",
       "UAP1|2_sgrna_chr1:162561996:162562016             2            UAP1  \n",
       "UBE2T|1_sgrna_chr1:202341918:202341938            1           UBE2T  \n",
       "UBE2T|2_sgrna_chr1:202341907:202341927            2           UBE2T  \n",
       "VDAC1|1_sgrna_chr5:134004687:134004707            1           VDAC1  \n",
       "VDAC1|2_sgrna_chr5:134004700:134004720            2           VDAC1  \n",
       "\n",
       "[98 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu_guides_and_rna['guides'].var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e434df3-1ccd-4403-a537-12f64534af13",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5109 × 16030\n",
       "    obs: 'bath_number', 'percent_mito', 'log_number_of_detected_genes', 'log_total_gene_count', 'log_total_guide_count'\n",
       "    var: 'feature_name', 'n_cells', 'transcript_chr', 'transcript_start', 'transcript_end', 'ensg'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu_guides_and_rna['scRNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "67f3e327-f585-49ff-8824-035ff3bd26bb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_name</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>transcript_chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>ensg</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>79</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.5</th>\n",
       "      <td>AL627309.5</td>\n",
       "      <td>417</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP006222.1</th>\n",
       "      <td>AP006222.1</td>\n",
       "      <td>240</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP006222.2</th>\n",
       "      <td>AP006222.2</td>\n",
       "      <td>51</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000286448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL732372.2</th>\n",
       "      <td>AL732372.2</td>\n",
       "      <td>90</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000237094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL592183.1</th>\n",
       "      <td>AL592183.1</td>\n",
       "      <td>250</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000273748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>AC007325.4</td>\n",
       "      <td>58</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000278817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL354822.1</th>\n",
       "      <td>AL354822.1</td>\n",
       "      <td>525</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000278384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004556.3</th>\n",
       "      <td>AC004556.3</td>\n",
       "      <td>3144</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000276345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>191</td>\n",
       "      <td>NOT_FOUND</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16030 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             feature_name  n_cells transcript_chr  transcript_start  \\\n",
       "feature_name                                                          \n",
       "AL627309.1     AL627309.1       79      NOT_FOUND                 0   \n",
       "AL627309.5     AL627309.5      417      NOT_FOUND                 0   \n",
       "AP006222.1     AP006222.1      240      NOT_FOUND                 0   \n",
       "AP006222.2     AP006222.2       51      NOT_FOUND                 0   \n",
       "AL732372.2     AL732372.2       90      NOT_FOUND                 0   \n",
       "...                   ...      ...            ...               ...   \n",
       "AL592183.1     AL592183.1      250      NOT_FOUND                 0   \n",
       "AC007325.4     AC007325.4       58      NOT_FOUND                 0   \n",
       "AL354822.1     AL354822.1      525      NOT_FOUND                 0   \n",
       "AC004556.3     AC004556.3     3144      NOT_FOUND                 0   \n",
       "AC240274.1     AC240274.1      191      NOT_FOUND                 0   \n",
       "\n",
       "              transcript_end             ensg  \n",
       "feature_name                                   \n",
       "AL627309.1                 0  ENSG00000238009  \n",
       "AL627309.5                 0  ENSG00000241860  \n",
       "AP006222.1                 0  ENSG00000228463  \n",
       "AP006222.2                 0  ENSG00000286448  \n",
       "AL732372.2                 0  ENSG00000237094  \n",
       "...                      ...              ...  \n",
       "AL592183.1                 0  ENSG00000273748  \n",
       "AC007325.4                 0  ENSG00000278817  \n",
       "AL354822.1                 0  ENSG00000278384  \n",
       "AC004556.3                 0  ENSG00000276345  \n",
       "AC240274.1                 0  ENSG00000271254  \n",
       "\n",
       "[16030 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu_guides_and_rna['scRNA'].var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "28013044-d38e-4c7a-aa74-79749fc43441",
   "metadata": {
    "scrolled": false,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bath_number</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>log_number_of_detected_genes</th>\n",
       "      <th>log_total_gene_count</th>\n",
       "      <th>log_total_guide_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGGCTT</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.513429</td>\n",
       "      <td>5.517453</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAATACAG</th>\n",
       "      <td>1</td>\n",
       "      <td>0.022816</td>\n",
       "      <td>8.161660</td>\n",
       "      <td>8.161946</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTCGACG</th>\n",
       "      <td>1</td>\n",
       "      <td>0.051611</td>\n",
       "      <td>7.784473</td>\n",
       "      <td>7.784889</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCTTCC</th>\n",
       "      <td>1</td>\n",
       "      <td>0.053218</td>\n",
       "      <td>7.934155</td>\n",
       "      <td>7.934513</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGCCTAA</th>\n",
       "      <td>1</td>\n",
       "      <td>0.045982</td>\n",
       "      <td>8.531490</td>\n",
       "      <td>8.531688</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCTGCAGG</th>\n",
       "      <td>1</td>\n",
       "      <td>0.027573</td>\n",
       "      <td>7.791110</td>\n",
       "      <td>7.791523</td>\n",
       "      <td>2.079442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCGTCCGTT</th>\n",
       "      <td>1</td>\n",
       "      <td>0.027464</td>\n",
       "      <td>8.423102</td>\n",
       "      <td>8.423322</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGCGAGAAA</th>\n",
       "      <td>1</td>\n",
       "      <td>0.040875</td>\n",
       "      <td>7.649693</td>\n",
       "      <td>7.650169</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACATAACC</th>\n",
       "      <td>1</td>\n",
       "      <td>0.031246</td>\n",
       "      <td>8.506739</td>\n",
       "      <td>8.506941</td>\n",
       "      <td>1.386294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACAGGGTACA</th>\n",
       "      <td>1</td>\n",
       "      <td>0.036355</td>\n",
       "      <td>8.281977</td>\n",
       "      <td>8.282230</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5109 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  bath_number  percent_mito  log_number_of_detected_genes  \\\n",
       "AAACCTGAGAGGGCTT            1      0.000000                      5.513429   \n",
       "AAACCTGCAAATACAG            1      0.022816                      8.161660   \n",
       "AAACCTGCACTCGACG            1      0.051611                      7.784473   \n",
       "AAACCTGCAGTCTTCC            1      0.053218                      7.934155   \n",
       "AAACCTGCATGCCTAA            1      0.045982                      8.531490   \n",
       "...                       ...           ...                           ...   \n",
       "TTTGGTTTCCTGCAGG            1      0.027573                      7.791110   \n",
       "TTTGGTTTCGTCCGTT            1      0.027464                      8.423102   \n",
       "TTTGTCAAGCGAGAAA            1      0.040875                      7.649693   \n",
       "TTTGTCACACATAACC            1      0.031246                      8.506739   \n",
       "TTTGTCACAGGGTACA            1      0.036355                      8.281977   \n",
       "\n",
       "                  log_total_gene_count  log_total_guide_count  \n",
       "AAACCTGAGAGGGCTT              5.517453               0.000000  \n",
       "AAACCTGCAAATACAG              8.161946               0.000000  \n",
       "AAACCTGCACTCGACG              7.784889               0.000000  \n",
       "AAACCTGCAGTCTTCC              7.934513               0.693147  \n",
       "AAACCTGCATGCCTAA              8.531688               0.000000  \n",
       "...                                ...                    ...  \n",
       "TTTGGTTTCCTGCAGG              7.791523               2.079442  \n",
       "TTTGGTTTCGTCCGTT              8.423322               0.693147  \n",
       "TTTGTCAAGCGAGAAA              7.650169               0.000000  \n",
       "TTTGTCACACATAACC              8.506941               1.386294  \n",
       "TTTGTCACAGGGTACA              8.282230               0.000000  \n",
       "\n",
       "[5109 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mu_guides_and_rna['scRNA'].obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b6d1180",
   "metadata": {},
   "source": [
    "# Creating a conda enviroment\n",
    "We're going to specify a conda environment to run our analysis, here's an example environment that uses scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0ba70321",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting perturb.yaml\n"
     ]
    }
   ],
   "source": [
    "%%writefile perturb.yaml\n",
    "name: perturb-conda-jamboree\n",
    "channels:\n",
    "    - conda-forge\n",
    "    - defaults\n",
    "dependencies:\n",
    "    - anndata\n",
    "    - scanpy\n",
    "    - mudata\n",
    "    - pandas >= 1.5.3\n",
    "    - session-info\n",
    "    - pip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a6ca1b21",
   "metadata": {
    "scrolled": false,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conda-forge/linux-64                                        Using cache\n",
      "conda-forge/noarch                                          Using cache\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "pkgs/main/linux-64 \u001b[90m━━━━━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\n",
      "pkgs/main/noarch   \u001b[90m━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\n",
      "pkgs/r/linux-64    \u001b[90m━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\n",
      "pkgs/r/noarch      \u001b[90m━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gpkgs/main/linux-64                                            No change\n",
      "pkgs/main/noarch                                              No change\n",
      "pkgs/r/noarch                                                 No change\n",
      "pkgs/r/linux-64                                               No change\n",
      "\u001b[?25h\n",
      "\n",
      "Looking for: ['anndata', 'scanpy', 'mudata', \"pandas[version='>=1.5.3']\", 'session-info', 'pip']\n",
      "\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/jovyan/.conda/envs/perturb-conda-jamboree\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - anndata\n",
      "   - scanpy\n",
      "   - mudata\n",
      "   - pandas[version='>=1.5.3']\n",
      "   - session-info\n",
      "   - pip\n",
      "\n",
      "\n",
      "  Package                    Version  Build                    Channel                    Size\n",
      "────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "  Install:\n",
      "────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[32m+ _libgcc_mutex     \u001b[0m           0.1  conda_forge              conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ _openmp_mutex     \u001b[0m           4.5  2_gnu                    conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ anndata           \u001b[0m         0.8.0  pyhd8ed1ab_1             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ brotli            \u001b[0m         1.0.9  h166bdaf_8               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ brotli-bin        \u001b[0m         1.0.9  h166bdaf_8               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ brotlipy          \u001b[0m         0.7.0  py310h5764c6d_1005       conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ bzip2             \u001b[0m         1.0.8  h7f98852_4               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ c-ares            \u001b[0m        1.18.1  h7f98852_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ ca-certificates   \u001b[0m     2022.12.7  ha878542_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cached-property   \u001b[0m         1.5.2  hd8ed1ab_1               conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cached_property   \u001b[0m         1.5.2  pyha770c72_1             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ certifi           \u001b[0m     2022.12.7  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cffi              \u001b[0m        1.15.1  py310h255011f_3          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ charset-normalizer\u001b[0m         3.1.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ colorama          \u001b[0m         0.4.6  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ contourpy         \u001b[0m         1.0.7  py310hdf3cbec_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cryptography      \u001b[0m        40.0.1  py310h34c0648_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cycler            \u001b[0m        0.11.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ fonttools         \u001b[0m        4.39.3  py310h1fa729e_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ freetype          \u001b[0m        2.12.1  hca18f0e_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ h5py              \u001b[0m         3.8.0  nompi_py310ha66b2ad_101  conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ hdf5              \u001b[0m        1.14.0  nompi_hb72d44e_103       conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ icu               \u001b[0m          72.1  hcb278e6_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ idna              \u001b[0m           3.4  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ importlib-metadata\u001b[0m         6.1.0  pyha770c72_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ importlib_metadata\u001b[0m         6.1.0  hd8ed1ab_0               conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ joblib            \u001b[0m         1.2.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ keyutils          \u001b[0m         1.6.1  h166bdaf_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ kiwisolver        \u001b[0m         1.4.4  py310hbf28c38_1          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ krb5              \u001b[0m        1.20.1  h81ceb04_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ lcms2             \u001b[0m          2.15  haa2dc70_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ ld_impl_linux-64  \u001b[0m          2.40  h41732ed_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ lerc              \u001b[0m         4.0.0  h27087fc_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libaec            \u001b[0m         1.0.6  hcb278e6_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libblas           \u001b[0m         3.9.0  16_linux64_openblas      conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libbrotlicommon   \u001b[0m         1.0.9  h166bdaf_8               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libbrotlidec      \u001b[0m         1.0.9  h166bdaf_8               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libbrotlienc      \u001b[0m         1.0.9  h166bdaf_8               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libcblas          \u001b[0m         3.9.0  16_linux64_openblas      conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libcurl           \u001b[0m        7.88.1  hdc1c0ab_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libdeflate        \u001b[0m          1.18  h0b41bf4_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libedit           \u001b[0m  3.1.20191231  he28a2e2_2               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libev             \u001b[0m          4.33  h516909a_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libffi            \u001b[0m         3.4.2  h7f98852_5               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgcc-ng         \u001b[0m        12.2.0  h65d4601_19              conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgfortran-ng    \u001b[0m        12.2.0  h69a702a_19              conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgfortran5      \u001b[0m        12.2.0  h337968e_19              conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgomp           \u001b[0m        12.2.0  h65d4601_19              conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libhwloc          \u001b[0m         2.9.0  hd6dc26d_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libiconv          \u001b[0m          1.17  h166bdaf_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libjpeg-turbo     \u001b[0m       2.1.5.1  h0b41bf4_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ liblapack         \u001b[0m         3.9.0  16_linux64_openblas      conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libllvm11         \u001b[0m        11.1.0  he0ac6c6_5               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libnghttp2        \u001b[0m        1.52.0  h61bc06f_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libnsl            \u001b[0m         2.0.0  h7f98852_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libopenblas       \u001b[0m        0.3.21  pthreads_h78a6416_3      conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libpng            \u001b[0m        1.6.39  h753d276_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libsqlite         \u001b[0m        3.40.0  h753d276_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libssh2           \u001b[0m        1.10.0  hf14f497_3               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libstdcxx-ng      \u001b[0m        12.2.0  h46fd767_19              conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libtiff           \u001b[0m         4.5.0  ha587672_6               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libuuid           \u001b[0m        2.38.1  h0b41bf4_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libwebp-base      \u001b[0m         1.3.0  h0b41bf4_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libxcb            \u001b[0m          1.13  h7f98852_1004            conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libxml2           \u001b[0m        2.10.3  hfdac1af_6               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libzlib           \u001b[0m        1.2.13  h166bdaf_4               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ llvmlite          \u001b[0m        0.39.1  py310h58363a5_1          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ matplotlib-base   \u001b[0m         3.7.1  py310he60537e_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ mudata            \u001b[0m         0.2.1  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ munkres           \u001b[0m         1.1.4  pyh9f0ad1d_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ natsort           \u001b[0m         8.3.1  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ ncurses           \u001b[0m           6.3  h27087fc_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ networkx          \u001b[0m           3.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ numba             \u001b[0m        0.56.4  py310ha5257ce_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ numpy             \u001b[0m        1.23.5  py310h53a5b5f_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ openjpeg          \u001b[0m         2.5.0  hfec8fc6_2               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ openssl           \u001b[0m         3.1.0  h0b41bf4_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ packaging         \u001b[0m          23.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pandas            \u001b[0m         1.5.3  py310h9b08913_1          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ patsy             \u001b[0m         0.5.3  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pillow            \u001b[0m         9.4.0  py310h065c6d2_2          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pip               \u001b[0m        23.0.1  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ platformdirs      \u001b[0m         3.2.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pooch             \u001b[0m         1.7.0  pyha770c72_3             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pthread-stubs     \u001b[0m           0.4  h36c2ea0_1001            conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pycparser         \u001b[0m          2.21  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pynndescent       \u001b[0m         0.5.8  pyh1a96a4e_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pyopenssl         \u001b[0m        23.1.1  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pyparsing         \u001b[0m         3.0.9  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pysocks           \u001b[0m         1.7.1  pyha2e5f31_6             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ python            \u001b[0m       3.10.10  he550d4f_0_cpython       conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ python-dateutil   \u001b[0m         2.8.2  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ python_abi        \u001b[0m          3.10  3_cp310                  conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pytz              \u001b[0m        2023.3  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ readline          \u001b[0m           8.2  h8228510_1               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ requests          \u001b[0m        2.28.2  pyhd8ed1ab_1             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ scanpy            \u001b[0m         1.9.3  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ scikit-learn      \u001b[0m         1.2.2  py310h41b6a48_1          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ scipy             \u001b[0m        1.10.1  py310h8deb116_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ seaborn           \u001b[0m        0.12.2  hd8ed1ab_0               conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ seaborn-base      \u001b[0m        0.12.2  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ session-info      \u001b[0m         1.0.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ setuptools        \u001b[0m        67.6.1  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ six               \u001b[0m        1.16.0  pyh6c4a22f_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ statsmodels       \u001b[0m        0.13.5  py310hde88566_2          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ stdlib-list       \u001b[0m         0.8.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ tbb               \u001b[0m      2021.8.0  hf52228f_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ threadpoolctl     \u001b[0m         3.1.0  pyh8a188c0_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ tk                \u001b[0m        8.6.12  h27826a3_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ tqdm              \u001b[0m        4.65.0  pyhd8ed1ab_1             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ typing-extensions \u001b[0m         4.5.0  hd8ed1ab_0               conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ typing_extensions \u001b[0m         4.5.0  pyha770c72_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ tzdata            \u001b[0m         2023c  h71feb2d_0               conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ umap-learn        \u001b[0m         0.5.3  py310hff52083_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ unicodedata2      \u001b[0m        15.0.0  py310h5764c6d_0          conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ urllib3           \u001b[0m       1.26.15  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ wheel             \u001b[0m        0.40.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xlrd              \u001b[0m         1.2.0  pyh9f0ad1d_1             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xorg-libxau       \u001b[0m         1.0.9  h7f98852_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xorg-libxdmcp     \u001b[0m         1.1.3  h7f98852_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xz                \u001b[0m         5.2.6  h166bdaf_0               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ zipp              \u001b[0m        3.15.0  pyhd8ed1ab_0             conda-forge/noarch  \u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ zstd              \u001b[0m         1.5.2  h3eb15da_6               conda-forge/linux-64\u001b[32m     Cached\u001b[0m\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 123 packages\n",
      "\n",
      "  Total download: 0 B\n",
      "\n",
      "────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\u001b[?25l\u001b[2K\u001b[0G\u001b[?25h\n",
      "Downloading and Extracting Packages\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate perturb-conda-jamboree\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mamba env create -f perturb.yaml --force"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "604ca621",
   "metadata": {
    "scrolled": false,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# conda environments:\r\n",
      "#\r\n",
      "perturb-conda-jamboree     /home/jovyan/.conda/envs/perturb-conda-jamboree\r\n",
      "perturbvi-conda-jamboree     /home/jovyan/.conda/envs/perturbvi-conda-jamboree\r\n",
      "base                     /opt/conda\r\n",
      "pygenomictracks          /opt/conda/envs/pygenomictracks\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!conda env list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7ffc7503",
   "metadata": {},
   "outputs": [],
   "source": [
    "!source activate perturb-conda-jamboree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c12039d3",
   "metadata": {},
   "source": [
    "# Developing a nextflow pipeline module"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ac8f14",
   "metadata": {},
   "source": [
    "## Nextflow job specification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b1fd4b3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting differential_pertubation.nf\n"
     ]
    }
   ],
   "source": [
    "%%writefile differential_pertubation.nf\n",
    "\n",
    "\n",
    "params.IN_TRANS = 'TRUE'\n",
    "params.MU_GUIDE_AND_SCRNA = '/home/jovyan/mudata_processed_guides_and_scrna.h5mu'\n",
    "\n",
    "\n",
    "\n",
    "useMamba = true\n",
    "//place to cash conda/mamba packages\n",
    "cacheDir =  '/home/jovyan/' \n",
    "\n",
    "process differential_pertubation {\n",
    "\n",
    "   debug true\n",
    "   conda '/home/jovyan/.conda/envs/perturb-conda-jamboree'\n",
    "    \n",
    "   input:\n",
    "   path (mudata_guides_rna)\n",
    "   val (in_trans)\n",
    "   output:\n",
    "   path 'results_dir', emit: differential_results_dir\n",
    "   script:\n",
    "    \"\"\"\n",
    "     mkdir -p results_dir\n",
    "     python /home/jovyan/differential_pertubation.py ${mudata_guides_rna} ${in_trans}\n",
    "    \"\"\"\n",
    "    \n",
    "}\n",
    "\n",
    "workflow {\n",
    "    differential_pertubation(params.MU_GUIDE_AND_SCRNA,\n",
    "                             params.IN_TRANS )\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5032ec79",
   "metadata": {},
   "source": [
    "## Python script for nextflow to run\n",
    "(You can also substitute in a R script or a wrapper around another tool, just change the extension and update the .nf file above)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3859d111",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting differential_pertubation.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile differential_pertubation.py\n",
    "import sys\n",
    "import os\n",
    "import mudata as md\n",
    "\n",
    "\n",
    "mu_guides_and_rna = md.read(sys.argv[1])\n",
    "in_trans = sys.argv[2]\n",
    "\n",
    "### do some analysis\n",
    "# ...\n",
    "\n",
    "#save all the results (or a mudata witht the results) in the results_dir dir. \n",
    "\n",
    "output_file = 'results_dir/'+os.path.basename(sys.argv[1]).replace('.h5mu','-final.h5mu')\n",
    "mu_guides_and_rna.write_h5mu(output_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "53e3c375",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N E X T F L O W  ~  version 22.10.6\n",
      "Launching `differential_pertubation.nf` [maniac_pesquet] DSL2 - revision: 64f37edfb1\n",
      "[-        ] process > differential_pertubation -\u001b[K\n",
      "\u001b[2A\n",
      "executor >  local (1)\u001b[K\n",
      "[76/e73dd8] process > differential_pertubation [  0%] 0 of 1\u001b[K\n",
      "\u001b[3A\n",
      "executor >  local (1)\u001b[K\n",
      "[76/e73dd8] process > differential_pertubation [100%] 1 of 1 ✔\u001b[K\n",
      "\u001b[K\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!nextflow run differential_pertubation.nf -w diff_perturb_test -with-conda"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3407071f",
   "metadata": {},
   "source": [
    "Your output data will show up in a subfolder of `diff_perturb_test`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "938639de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 52\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 15:29 \u001b[0m\u001b[01;34m76\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 15:20 \u001b[01;34m1b\u001b[0m/\r\n",
      "drwxr-sr-x 4 jovyan users 4096 Apr  3 15:16 \u001b[01;34m39\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 15:14 \u001b[01;34m3d\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 15:13 \u001b[01;34m31\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 15:10 \u001b[01;34m10\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 15:08 \u001b[01;34me9\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 14:53 \u001b[01;34m37\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 14:51 \u001b[01;34md7\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 14:45 \u001b[01;34m21\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 14:43 \u001b[01;34md5\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 14:17 \u001b[01;34mf8\u001b[0m/\r\n",
      "drwxr-sr-x 3 jovyan users 4096 Apr  3 14:17 \u001b[01;34me2\u001b[0m/\r\n"
     ]
    }
   ],
   "source": [
    "ls -lt diff_perturb_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be412279",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
